[JSC] Refactor Wasm::Table and fix memory leak

Constellation opened the pull request:
#### 55bd5e05561eb0c7c119fce0c90d5c1c1b24fab9
<pre>
[JSC] Refactor Wasm::Table and fix memory leak
<a href="https://bugs.webkit.org/show_bug.cgi?id=249861">https://bugs.webkit.org/show_bug.cgi?id=249861</a>
rdar://103682266

Reviewed by Mark Lam.

This patch refactors Wasm::Table and fixes memory leak.
Previously, FuncRefTable&apos;s destructor is not called so
it was leaking memory for functions. This patch integrates
std::destroying_delete so that we can call derived class&apos;
destructor correctly. We also introduce ExternRefTable so
we allocate combined memory (jsValue + functions) for FuncRefTable.

* Source/JavaScriptCore/runtime/WriteBarrier.h:
* Source/JavaScriptCore/wasm/WasmFormat.h:
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::setWasmTableElement):
* Source/JavaScriptCore/wasm/WasmTable.cpp:
(JSC::Wasm::Table::visitDerived):
(JSC::Wasm::Table::visitDerived const):
(JSC::Wasm::Table::operator delete):
(JSC::Wasm::Table::Table):
(JSC::Wasm::Table::tryCreate):
(JSC::Wasm::Table::grow):
(JSC::Wasm::Table::clear):
(JSC::Wasm::Table::set):
(JSC::Wasm::Table::get const):
(JSC::Wasm::Table::visitAggregateImpl):
(JSC::Wasm::ExternRefTable::ExternRefTable):
(JSC::Wasm::ExternRefTable::clear):
(JSC::Wasm::ExternRefTable::set):
(JSC::Wasm::FuncRefTable::FuncRefTable):
(JSC::Wasm::FuncRefTable::setFunction):
(JSC::Wasm::FuncRefTable::copyFunction):
(JSC::Wasm::FuncRefTable::clear):
(JSC::Wasm::FuncRefTable::set):
* Source/JavaScriptCore/wasm/WasmTable.h:
(JSC::Wasm::FuncRefTable::Function::offsetOfFunction): Deleted.
(JSC::Wasm::FuncRefTable::Function::offsetOfInstance): Deleted.
(JSC::Wasm::FuncRefTable::offsetOfFunctions): Deleted.
* Source/JavaScriptCore/wasm/js/JSWebAssemblyTable.cpp:
(JSC::JSWebAssemblyTable::set):

Canonical link: <a href="https://commits.webkit.org/258325@main">https://commits.webkit.org/258325@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7d13390f4d5212cf20955f9508e6da74d19128b0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101522 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10677 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34577 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110809 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/171050 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11632 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1576 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93915 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108610 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107303 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8858 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/92100 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/35385 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90752 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/23513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/78383 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/91942 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4276 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/25018 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/88092 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/1865 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4342 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/1475 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29343 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10427 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/44506 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/90994 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6104 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20336 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3014 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/wasm/WasmTable.cpp:
```diff
@@ -196,17 +257,20 @@ FuncRefTable::FuncRefTable(uint32_t initial, std::optional<uint32_t> maximum)
-    m_importableFunctions.get()[index] = function;
+        m_importableFunctions.get()[index].m_value.set(m_owner->vm(), m_owner, optionalWrapper);
+    else
+        m_importableFunctions.get()[index].m_value.set(m_owner->vm(), m_owner, jsNull());
```
Is it worth pre-caching `m_importableFunctions.get()[index]` in a local?

Secondly, why not use a `?:` expression (instead of the explicit `if` statement) for setting `m_value`?

Constellation replied:
Also, I think probably the second one can use setEarlyValue. Changed.

MenloDorian reviewed (approved):
r=me

Constellation committed:
[JSC] Refactor Wasm::Table and fix memory leak
https://bugs.webkit.org/show_bug.cgi?id=249861
rdar://103682266

Reviewed by Mark Lam.

This patch refactors Wasm::Table and fixes memory leak.
Previously, FuncRefTable's destructor is not called so
it was leaking memory for functions. This patch integrates
std::destroying_delete so that we can call derived class'
destructor correctly. We also introduce ExternRefTable so
we allocate combined memory (jsValue + functions) for FuncRefTable.

* Source/JavaScriptCore/runtime/WriteBarrier.h:
* Source/JavaScriptCore/wasm/WasmFormat.h:
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::setWasmTableElement):
* Source/JavaScriptCore/wasm/WasmTable.cpp:
(JSC::Wasm::Table::visitDerived):
(JSC::Wasm::Table::visitDerived const):
(JSC::Wasm::Table::operator delete):
(JSC::Wasm::Table::Table):
(JSC::Wasm::Table::tryCreate):
(JSC::Wasm::Table::grow):
(JSC::Wasm::Table::clear):
(JSC::Wasm::Table::set):
(JSC::Wasm::Table::get const):
(JSC::Wasm::Table::visitAggregateImpl):
(JSC::Wasm::ExternRefTable::ExternRefTable):
(JSC::Wasm::ExternRefTable::clear):
(JSC::Wasm::ExternRefTable::set):
(JSC::Wasm::FuncRefTable::FuncRefTable):
(JSC::Wasm::FuncRefTable::setFunction):
(JSC::Wasm::FuncRefTable::copyFunction):
(JSC::Wasm::FuncRefTable::clear):
(JSC::Wasm::FuncRefTable::set):
* Source/JavaScriptCore/wasm/WasmTable.h:
(JSC::Wasm::FuncRefTable::Function::offsetOfFunction): Deleted.
(JSC::Wasm::FuncRefTable::Function::offsetOfInstance): Deleted.
(JSC::Wasm::FuncRefTable::offsetOfFunctions): Deleted.
* Source/JavaScriptCore/wasm/js/JSWebAssemblyTable.cpp:
(JSC::JSWebAssemblyTable::set):

Canonical link: https://commits.webkit.org/258325@main
